src/lib.rs

```diff
@@ -46,6 +67,200 @@ impl UniffiCustomTypeConverter for Uuid {
+        value: e.to_string(),
+    })?;
+    let dl: DlSubfile = file
+        .read_subfile(b"DL")
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .ok_or(VCBVerificationError::Generic {
+            value: "Invalid DLSubfile".to_string(),
+        })?;
+    let zz: ZZSubfile = file
+        .read_subfile(b"ZZ")
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .ok_or(VCBVerificationError::Generic {
+            value: "Invalid ZZSubfile".to_string(),
+        })?;
+    let vc = zz
+        .decode_credential()
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?;
+
+    let status_list_client = ConstTerseStatusListProvider::new(
+        StatusLists,
+        StatusListInfo::new(1000, StatusPurpose::Revocation),
+    );
+
+    let params = VerificationParameters::new_with(
+        AnyDidMethod::default().into_vm_resolver(),
+        status_list_client,
+    );
+
+    verify(&vc, &dl.mandatory, params)
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })
+}
+
+fn convert_to_mrz_entry(s: &[u8]) -> &[u8; 30] {
+    s.try_into().expect("slice with incorrect length")
+}
+
+#[derive(thiserror::Error, uniffi::Error, Debug)]
+pub enum VCBVerificationError {
+    #[error("{value}")]
+    Generic { value: String },
+}
+
+#[uniffi::export]
+pub async fn verify_vcb_qrcode_against_mrz(
+    mrz_payload: String,
+    qr_payload: String,
+) -> Result<(), VCBVerificationError> {
+    let mrz: MRZ = mrz_payload
+        .lines()
+        .map(|x| *convert_to_mrz_entry(x.as_bytes()))
+        .collect::<Vec<[u8; 30]>>()
+        .try_into()
+        .map_err(|_| VCBVerificationError::Generic {
+            value: "Invalid MRZ string".to_string(),
+        })?;
+
+    // First we decode the QR-code payload to get the VCB in CBOR-LD form.
+    let input = MachineReadableZone::decode_qr_code_payload(qr_payload.as_str()).map_err(|e| {
+        VCBVerificationError::Generic {
+            value: e.to_string(),
+        }
+    })?;
+
+    // Then we decompress the CBOR-LD VCB to get a regular JSON-LD VCB.
+    let vc = decode_from_bytes::<MachineReadableZone>(&input)
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?;
+
+    // Finally we verify the VCB against the MRZ data.
+    let params = VerificationParameters::new(AnyDidMethod::default().into_vm_resolver());
+    verify(&vc, &mrz, params)
+        .await
+        .map_err(|e| VCBVerificationError::Generic {
+            value: e.to_string(),
+        })?
+        .map_err(|e| VCBVerificationError::Generic {
```

Comment:
Could you add a new variant to say "verification failed" -- so that the app can distinguish between something going wrong and the verification simply failing